### ISSUE TITLE
Hash fragment not cleared

### ISSUE THREAD

**Issue by MaciejWierzchowski** [jeroenheijmans/sample-angular-oauth2-oidc-with-auth-guards#8]
Hi,

As a background, I'm using this example as my starting point, and have made a few changes. The biggest one (to the flow itself) is this:
* At step 3, i use login method. Basically, if silentRefresh fails, I want to immidiately redirect to Identity Server. I'm calling the login method, so that I can be redirected to the correct page that user initially navigated to. 

https://github.com/jeroenheijmans/sample-angular-oauth2-oidc-with-auth-guards/blob/bdc4d55eab04c62b8b94806c4ddd89eaf200a02a/src/app/auth.service.ts#L127-L128

But since I am also using guards, this of course wont work (`router.url` doesn't have the same url as in address bar, since the guard is blocking navigation).

https://github.com/jeroenheijmans/sample-angular-oauth2-oidc-with-auth-guards/blob/bdc4d55eab04c62b8b94806c4ddd89eaf200a02a/src/app/auth.service.ts#L156

In any case, the problem I have noticed, is that once I get back after logging in, the hash fragment is not actually cleared by the angular-oauth2-oidc (or maybe it is but then is readded, since it does visually blink). 

The example here *seems* to be working (i.e. the hash is not there) only because of the navigation:
https://github.com/jeroenheijmans/sample-angular-oauth2-oidc-with-auth-guards/blob/bdc4d55eab04c62b8b94806c4ddd89eaf200a02a/src/app/auth.service.ts#L149

Do you have slightest idea? I know it's a bit vague description, I'm just asking if this is something you have seen.

**Comment by jeroenheijmans** [comment]
I might not have a lot of time the coming weeks to respond, but here's a quick one FWIW for now.

In your first bullet, you mention that you've enabled the "ALWAYS force login" feature. I presume from your text that you do try to send some kind of target url (encoded) along, but that it's the "wrong" one because guards might still be blocking things.

I'm not entirely sure how I'd handle such a case, and would have to try things, but some thoughts:

- Can we send just the `location.href` or some other part of the `location`? The user probably used a bookmark or other deeplink, right?
- Can we inspect the `Router` or its `State` to see what the _intended_ route is?
- As a fallback/workaround, you could have a generic guard that does things, so just protect *all* routes with [`auth-guard-with-forced-login`](https://github.com/jeroenheijmans/sample-angular-oauth2-oidc-with-auth-guards/blob/master/src/app/auth-guard-with-forced-login.service.ts)?

Hope that helps. Let me know if you find a solution or workaround.

**Comment by MaciejWierzchowski** [comment]
Yeah, I pretty much started with Router.state, but couldn't find anything of value there. But then, it might be possible to find the correct url in the guard... That's something I will look into.

As to your last question: Sorry for that, this is new to me.
My point from header is that while I was playing with the ideas above (e.g. using location.href etc.) for preserving state, I found a bit unexpected behaviour: That the *hash fragment* is being removed by this line:
https://github.com/jeroenheijmans/sample-angular-oauth2-oidc-with-auth-guards/blob/bdc4d55eab04c62b8b94806c4ddd89eaf200a02a/src/app/auth.service.ts#L149

While I would expect it to be removed here:
https://github.com/jeroenheijmans/sample-angular-oauth2-oidc-with-auth-guards/blob/bdc4d55eab04c62b8b94806c4ddd89eaf200a02a/src/app/auth.service.ts#L94

Again, I do have some differences between what's in this repo and in my code. In this case, I use `loadDiscoveryDocumentAndTryLogin()`.

**Comment by MaciejWierzchowski** [comment]
Ok, after some debugging, I think I've figured this out. 
The problem lies here:
https://github.com/jeroenheijmans/sample-angular-oauth2-oidc-with-auth-guards/blob/bdc4d55eab04c62b8b94806c4ddd89eaf200a02a/src/app/auth.service.ts#L142

This line basically resolves the `canActivateProtectedRoutes$` to true. The guard has a subscription on this, and as soon as it happens, it allows the navigation. But after the navigation back, the navigation contains the hash fragment. 

So basically:
1. Redirect back from IS
2. LoadDocumentAndTryLogin(). This removes hash.
3. Token is received, so `isAuthenticated$` is true. 
4. Posting a value to `isDoneLoadingSubject$` allows the navigation to the pending route, which is the one with hash.

What I did was move the `isDoneLoadingSubject$` to the end of that handler (`after the router.navigateByUrl`)
